algo/src/main/java/com/hartwig/serve/sources/ckb/ActionableTrialFactory.java

```diff
@@ -123,16 +127,18 @@ && hasPotentiallyOpenRequirementToInclude(trial.recruitment()) && hasAgeGroupToI
 
     @NotNull
     @VisibleForTesting
-    static Set<String> filterOnRegionsToInclude(@NotNull ClinicalTrial trial, @NotNull Set<CkbRegion> regionsToInclude) {
-        Set<String> countries = Sets.newHashSet();
-        for (Location location : trial.locations()) {
-            if (regionsToInclude.stream().anyMatch(region -> region.includes(location))
-                    && hasPotentiallyOpenRequirementToInclude(location.status())) {
-                countries.add(location.country());
-            }
-        }
-
-        return countries;
+    static Set<Country> extractCountriesToInclude(@NotNull ClinicalTrial trial, @NotNull Set<CkbRegion> regionsToInclude) {
+        Map<String, Map<String, Set<String>>> countriesToCitiesToHospitalNames = trial.locations()
+                .stream()
+                .filter(location -> regionsToInclude.stream().anyMatch(region -> region.includes(location))
+                        && hasPotentiallyOpenRequirementToInclude(location.status()))
+                .collect(Collectors.groupingBy(Location::country,
+                        Collectors.groupingBy(Location::city, Collectors.mapping(Location::facility, Collectors.toSet()))));
```

Comment:
I suspect this may even impress John :)

Reply:
haha I will be honest, Mr Chat Geepeetee helped me a bit here